src/sds.c

```diff
@@ -124,8 +132,14 @@ sds _sdsnewlen(const void *init, size_t initlen, int trymalloc) {
     s = (char*)sh+hdrlen;
     fp = ((unsigned char*)s)-1;
     usable = usable-hdrlen-1;
+
+#ifdef USE_JEMALLOC
+    assert(usable <= sdsTypeMaxSize(type));
```

Comment:
The assertion here is based on the behavior we expected/monitor from jemalloc, it depends on the internal of jemalloc, is there any guarantee from jemalloc? Maybe a potential risk.

Reply:
why can't we always keep the defending part?  (usable = min(usable, sdsTypeMaxSize(type))
AND also maybe we need to consider keep the assert for the usable size correctness, eg:
assert(usable >= initlen);

Reply:
```
assert(usable <= sdsTypeMaxSize(type));
assert(initlen <= usable);
```
Are our two invariants. Should we be doing those asserts regardless of the allocator being used?

Reply:
Yes that is a good point:
we want to make sure that in case JEMALLOC is used we will SURELY have **real** usable == usable, in order to satisfy #453, so I guess we do have to keep assert(usable <= sdsTypeMaxSize(type)); for JEMALLOC (for example in order to assert when JEMALLOC was compiled with different bin sizes?).
I think (as @madolson suggested) the defensive option is better at this point.

Reply:
I would still change the code so that:

> #ifndef USE_JEMALLOC
> usable = MIN(usable, sdsTypeMaxSize(type));
> #endif
> assert(usable <= sdsTypeMaxSize(type) && initlen <= usable);

Reply:
I think there is a better solution. We can adjust types after the allocation/reallocation. Previously I thought there are some edge cases that wouldn't work. But after reconsidering the idea, and noting again that `alloc` doesn't include header size and null terminator, I think it's a better way.
With this solution we don't need to take assumptions for correctness. Instead of truncating `usable`, we're going to adjust the header type.
I'll update the PR with this idea.